### PR TITLE
Use ctime in file digest cache key

### LINKED ISSUE #14723: Action outputs with fixed timestamps cause problems
### Description of the problem / feature request:

If an output file from a Bazel action has a fixed timestamp then if the content of the file change but the size of the file doesn't it won't trigger the re-running of any actions that depends on this file.

### Feature requests: what underlying problem are you trying to solve with this feature?

Some of the inputs to our build are archives containing template files that are expanded during the build. The archives are produced by another build system and to ensure that build system is reproducible the timestamps of the template files are set to 1970 in the archive. During the build we extract the templates files from the archive (retaining their timestamp as this is the default behavior of tar), expand the template and then use the expanded template in a further action. When we checkout a different refpoint of our code sometimes the archive has changed but the templates are still the same size and this results in actions not being re-run when they should be.

I understand that Bazel can't rehash all input files but I was a bit surprised that when an action is run that all output files are not rehashed. Perhaps this is too much of an edge case to worry about but I thought it was worth raising.

We've worked around the issue by not retaining timestamps when we extract files from the archives but wanted to check if Bazel should handle this.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Create a workspace with the following BUILD and .bzl files

```
$cat BUILD
load("rules.bzl", "my_expand")

genrule(
    name = "my_templates",
    srcs = ["template_archive.tar"],
    outs = ["template1"],
    cmd = "tar -C $(@D) -xf template_archive.tar",
)

my_expand(
    name = "expand1",
    input = "template1",
    output = "expanded1",
    to_sub = {"test":"foo"}
)
```

```
$cat rules.bzl
def _my_expand_impl(ctx):
    ctx.actions.expand_template(
        template = ctx.file.input,
        output = ctx.outputs.output,
        substitutions = ctx.attr.to_sub
    )

my_expand = rule(
    implementation = _my_expand_impl,
    attrs = {
        "input": attr.label(allow_single_file=True),
        "output": attr.output(),
        "to_sub" : attr.string_dict(),
    }
)
```

Now create a set of archives with different template files
```
echo "test : alpha" > template1
tar -cf template_archive_alpha.tar --mtime='1970-01-01' template1
echo "test : delta" > template1
tar -cf template_archive_delta.tar --mtime='1970-01-01' template1
echo "test : beta" > template1
tar -cf template_archive_beta.tar --mtime='1970-01-01' template1
```

Put the alpha archive in place as the input, build and verify that the expanded template is as expected.
```
$cp template_archive_alpha.tar template_archive.tar 
$bazel ...
INFO: Analyzed 2 targets (0 packages loaded, 0 targets configured).
INFO: Found 2 targets...
INFO: Elapsed time: 0.106s, Critical Path: 0.03s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
$cat bazel-bin/expanded1 
foo : alpha
```

Switch to the delta archive as the input, build and verify that the expanded template has not been updated.
```
$cp template_archive_delta.tar template_archive.tar 
$bazel ...
INFO: Analyzed 2 targets (0 packages loaded, 0 targets configured).
INFO: Found 2 targets...
INFO: Elapsed time: 0.095s, Critical Path: 0.03s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
$cat bazel-bin/expanded1 
foo : alpha
```

Even though the extracted template has been updated
```
$cat bazel-bin/template1 
test : delta
```

Verify that performing a Bazel clean and building results in the correct output.
```
$bazel clean
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
$bazel build ...
INFO: Analyzed 2 targets (5 packages loaded, 10 targets configured).
INFO: Found 2 targets...
INFO: Elapsed time: 0.186s, Critical Path: 0.03s
INFO: 3 processes: 2 internal, 1 linux-sandbox.
INFO: Build completed successfully, 3 total actions
$cat bazel-bin/expanded1 
foo : delta
```

Switching to/from the alpha/delta archives and the beta archive works fine presumably because the size of the template files change (beta being 4 characters and alpha/delta being 5 characters).

### What operating system are you running Bazel on?
Linux

### What's the output of `bazel info release`?

release 4.1.0- (@non-git)

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

I don't have full details to hand about how we built Bazel but I'm confident that any changes have not impacted the behavior described above.

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

N/A

###  Have you found anything relevant by searching the web?

No. I'm not even confident if Bazel should handle this scenario but I think it's worth asking.

### Any other information, logs, or outputs that you want to share?

No

### PR DESCRIPTION
**Pull request by fmeum**
File digests are now additionally keyed by ctime for supported file system implementations. Since Bazel has a non-zero default for `--cache_computed_file_digests`, this may be required for correctness in cases where different files have identical mtime and inode number. For example, this can happen on Linux when files are extracted from a tar file with fixed mtime and are then replaced with `mv`, which preserves inodes.

Since Java (N)IO doesn't have support for reading file ctimes on Windows, a new method backed by a native implementation is added to `WindowsFileOperation`. Adding a call to this function to `stat` uncovered previously silent bugs where Unix-style `PathFragment`s were created on Windows:

1. Bzlmod's `createLocalRepoSpec` did not correctly extract the path from a registry's `file://` URI on Windows.
2. `--package_path` isn't usable with absolute paths on Windows as it splits on `:`. Since the flag is deprecated, this commit fixes the tests rather than the implementation.

Fixes #14723

### CODE REVIEW

**Comment by fmeum** [comment]
@janakdr Could you review this fix?

`test_cache_computed_file_digests_behavior` is failing as it incurs two cache misses rather than getting the expected cache hit for `stable-status.txt`. This is surprising to me as `BazelWorkspaceStatusAction` only recreates `stable-status.txt` if the content changed, which it shouldn't in this test.

**Comment by justinhorvitz** [comment]
> test_cache_computed_file_digests_behavior is failing as it incurs two cache misses rather than getting the expected cache hit for stable-status.txt. This is surprising to me as BazelWorkspaceStatusAction only recreates stable-status.txt if the content changed, which it shouldn't in this test.

Does the latest change in https://github.com/bazelbuild/bazel/pull/18003/commits/1affa3e6eec862e55e742cd0b8a2a23fb283d6d7 fix this?

**Comment by fmeum** [comment]
@justinhorvitz Yes, at least as far as the tests are concerned. I don't know whether the cache thrashing is only relevant for `stable-status.txt` because it's generating actions runs on every build or whether this would affect performance more broadly. If it could, then tree artifact files may still be affected.

Regarding ctime on Windows: I could add native functions that get it, but this requires an additional stat call per file. I have no intuition for performance on Windows, would that still be better than disabling the digest cache?

**Review by justinhorvitz — CHANGES_REQUESTED** [review]

**Comment by justinhorvitz** [comment]
> @justinhorvitz Yes, at least as far as the tests are concerned. I don't know whether the cache thrashing is only relevant for `stable-status.txt` because it's generating actions runs on every build or whether this would affect performance more broadly. If it could, then tree artifact files may still be affected.
> 
> Regarding ctime on Windows: I could add native functions that get it, but this requires an additional stat call per file. I have no intuition for performance on Windows, would that still be better than disabling the digest cache?

I don't have much personal stake in external bazel or Windows performance so let's ask someone else. @tjgq do you see any potential scenario where tree artifact outputs would have a changed `ctime` but the same `mtime`?

**Comment by fmeum** [comment]
> I don't have much personal stake in external bazel or Windows performance so let's ask someone else. @tjgq do you see any potential scenario where tree artifact outputs would have a changed `ctime` but the same `mtime`?

Just want to note that these are separate issues:
1. `ActionMetadataHandler` may still thrash the digest cache for tree artifacts by unconditionally chmodding them on all OSes.
2. Windows doesn't provide an actual ctime, which means that this PR can't fix the actual issue.

**Comment by justinhorvitz** [comment]
> > I don't have much personal stake in external bazel or Windows performance so let's ask someone else. @tjgq do you see any potential scenario where tree artifact outputs would have a changed `ctime` but the same `mtime`?
> 
> Just want to note that these are separate issues:
> 
> 1. `ActionMetadataHandler` may still thrash the digest cache for tree artifacts by unconditionally chmodding them on all OSes.
> 2. Windows doesn't provide an actual ctime, which means that this PR can't fix the actual issue.

Yup, understood.

File: src/main/java/com/google/devtools/build/lib/vfs/FileStatus.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/FileStatus.java:97]: Thanks for the pointer, I implemented the new method for `InMemoryContentInfo` and also expanded the Javadoc.

File: src/main/java/com/google/devtools/build/lib/vfs/inmemoryfs/InMemoryContentInfo.java
Comment (justinhorvitz) [inline src/main/java/com/google/devtools/build/lib/vfs/inmemoryfs/InMemoryContentInfo.java:124]: I don't see what this needs to be synchronized with. Also, please make it `final`.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/inmemoryfs/InMemoryContentInfo.java]: Made it `final` and removed `synchronized`. I am not entirely sure about the intended behavior of `getPermissions` when the file permissions are updated concurrently, but I agree that `synchronized` doesn't help as the setters aren't synchronized.

**Review by justinhorvitz — APPROVED** [review]
This PR looks good to me, but I think we should get a second opinion specifically for the tree artifact question.

File: src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java
Comment (justinhorvitz) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java:68]: This idea seems sound under typical circumstances. I would just be concerned about use cases that play tricks with these timestamps. My general advice would be to not risk it unless there’s a tangible concern (e.g. in this case the increased memory use of the cache key, although if that’s a concern we could optimize for the common case of mtime == ctime by sharing a field).

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java:68]: Is this cached used at Google? If so I would be much less worried about this causing an unintentional memory regression. Since the cache key is critical for correctness, I would also prefer to leave it as is.

Comment (justinhorvitz) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java:68]: It’s not used under typical circumstances at google since fast digests are available.

Comment (justinhorvitz) [reply src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java:68]: Anyway, there’s a maximum size configured on the cache which can be adjusted, so I do not think you have to worry at all about the size if the keys, I was just giving an example of why we might want to pursue a riskier change.

**Comment by fmeum** [comment]
@meteorcloudy Could you take a look at the two latest commits? The first one adds support for ctimes to `WindowsFileSystem`, the next one fixes two issues that adding this support uncovered: In certain cases, Unix-style `PathFragment` instances were created on Windows, which fail when passed to the native Windows file system operations. I think that such errors have been silent so far as e.g. `IsSymlinkOrJunction` falls back to Java IO when the native function fails.

**Comment by tjgq** [comment]
I don't have any concerns regarding tree artifacts.

**Review by meteorcloudy — APPROVED** [review]
The Windows changes look good to me! Thanks!

**Comment by meteorcloudy** [comment]
Please resolve the conflict then we can import this PR.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by keertk** [comment]
@bazel-io fork 6.2.0

**Comment by fmeum** [comment]
@keertk @meteorcloudy I think that we should probably fork this for 5.4.1 and 6.1.2 as well as it fixes a correctness issue.

**Comment by keertk** [comment]
@bazel-io fork 5.4.1

**Comment by keertk** [comment]
@bazel-io fork 6.1.2

**Comment by meteorcloudy** [comment]
@fmeum Some update, we are importing this CL, but one internal test was broken, looking into a fix.

File: src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java
Comment (justinhorvitz) [inline src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java:730]: Should this use `statIfFound`? All of the other stats in this file use that to tolerate a nonexistent file.

**Comment by justinhorvitz** [comment]
> @fmeum Some update: we are importing this CL, but one internal test was broken, looking into a fix.

The broken test was counting the number of stats, but wasn't doing so in a very robust way. I've updated it which should unblock the import. Meanwhile, while looking into this, I found I have one more question which I asked in `ActionMetadataHandler`.

File: src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java]: True, it should to preserve the original semantics. I pushed a commit to do this. I am a bit surprised that no test failed due to this change - do all missing files at this point result in a `missing outputs` error down the line anyway?

**Comment by ulrfa** [comment]
I'm thinking about https://bazel.build/reference/command-line-reference#flag--experimental_use_hermetic_linux_sandbox that populates sandboxes by creating hard links.

Ctime is changed when additional hard links are created to existing inodes (link count changes).

How is this change affecting the file digest cache for that use case?

**Comment by fmeum** [comment]
> I'm thinking about https://bazel.build/reference/command-line-reference#flag--experimental_use_hermetic_linux_sandbox that populates sandboxes by creating hard links.
> 
> Ctime is changed when additional hard links are created to existing inodes (link count changes).
> 
> How is this change affecting the file digest cache for that use case?

I assume it would make it so that digests are recomputed every time a hard link is created, which doesn't sound good. Could you maybe try this out with Bazel built from this PR?

I don't know enough about the hermetic sandbox to think of ways to improve digest caching for it. Since Bazel doesn't really care about changes to metadata, we would ideally like to have a "modification time, but not overrideable by users", for which `ctime` is really just a proxy, but I don't see a way to get this more restricted information efficiently.

**Pull request closed without merging**
